pkg/gateway/gateway_test.go

```diff
@@ -4,11 +4,177 @@ package gateway
+		},
+		{
+			name:           "GatewayDomains are split, formatted, and added to allowList",
+			gatewayDomains: "doMain1,Domain2",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(emptyEnv)
+			},
+			expectedAllowList: []string{"domain1", "domain2"},
+		},
+		{
+			name: "Domains in Environments are parsed and appended to allowList",
+			mocks: func(env *mock_env.MockCore) {
+				populatedEnv := *emptyEnv
+				populatedEnv.ActiveDirectoryEndpoint = "https://activedirectory/redhattest"
+				populatedEnv.ResourceManagerEndpoint = "https://resourcemanager/redhattest"
+				env.EXPECT().Environment().AnyTimes().AnyTimes().Return(&populatedEnv)
+			},
+			expectedAllowList: []string{
+				"activedirectory",
+				"resourcemanager",
+			},
+		},
+		{
+			name: "Invalid domain format in environment",
+			mocks: func(env *mock_env.MockCore) {
+				invalidEnv := *emptyEnv
+				invalidEnv.ActiveDirectoryEndpoint = "\r"
+				env.EXPECT().Environment().AnyTimes().AnyTimes().Return(&invalidEnv)
+			},
+			wantErr: errors.New(`parse "\r": net/url: invalid control character in URL`),
+		},
+		{
+			name:          "When the acrResourceID is present it is parsed into domains and added to allowList",
+			acrResourceID: "/subscriptions/93aeba23-2f76-4307-be82-02921df010cf/resourceGroups/global/providers/Microsoft.ContainerRegistry/registries/resourceName",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(emptyEnv)
+				env.EXPECT().Location().AnyTimes().Return("location")
+			},
+			expectedAllowList: []string{
+				"resourcename.dnssuffix",
+				"resourcename.location.data.dnssuffix",
+			},
+		},
+		{
+			name:          "When the acrResourceID is present but it is invalid",
+			acrResourceID: "invalidAcrResourceID",
+			mocks: func(env *mock_env.MockCore) {
+				env.EXPECT().Environment().AnyTimes().Return(emptyEnv)
+				env.EXPECT().Location().AnyTimes().Return("location")
+			},
+			wantErr: errors.New("parsing failed for invalidAcrResourceID. Invalid resource Id format"),
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			controller := gomock.NewController(t)
+			defer controller.Finish()
+
+			baseLog := logrus.NewEntry(logrus.StandardLogger())
+			db := mock_database.NewMockGateway(controller)
+			metrics := mock_metrics.NewMockEmitter(controller)
+
+			httpl, _ := utilnet.Listen("tcp", ":8080", SocketSize)
+			httpsl, _ := utilnet.Listen("tcp", ":8443", SocketSize)
+
+			env := mock_env.NewMockCore(controller)
+			tt.mocks(env)
+
+			gtwy, err := NewGateway(ctx, env, baseLog, baseLog, db, httpsl, httpl, tt.acrResourceID, tt.gatewayDomains, metrics)
+
+			if err != nil {
+				if tt.wantErr == nil {
+					t.Errorf("Got an unexpected error: %s", err.Error())
+				}
+				if tt.wantErr.Error() != err.Error() {
+					t.Errorf("The expected error: %s did not match the actual error: %s", tt.wantErr.Error(), err.Error())
+				}
+				return
+			}
```

Comment:
This doesn't cover the case where we don't get an error but expect one.  I suggest that you change the type of `wantErr` to a `string` and do soemthing like:

```suggestion
			gtwy, err := NewGateway(ctx, env, baseLog, baseLog, db, httpsl, httpl, tt.acrResourceID, tt.gatewayDomains, metrics)
				if err != nil && err.Error() != tt.wantErr {
				t.Errorf("got error '%v', wanted error '%v'", err, tt.wantErr)
			}

			if err == nil && tt.wantErr != "" {
				t.Errorf("did not get an error, but wanted error '%v'", tt.wantErr)
			}
```

Reply:
forgive the indentation issues here.